pfl/privacy/joint_mechanism.py

```diff
@@ -0,0 +1,128 @@
+# Copyright © 2023-2024 Apple Inc.
+'''
+Joint mechanism for combining multiple mechanisms into one.
+'''
+
+from typing import Dict, List, Optional, Set, Tuple
+
+from pfl.metrics import Metrics, StringMetricName
+from pfl.stats import MappedVectorStatistics, TrainingStatistics
+
+from .privacy_mechanism import CentrallyApplicablePrivacyMechanism
+
+
+def check_if_partition(full_set: Set[str], partition: List[Set[str]]):
+    """
+    Checks if each element of full_list appears in exactly one of
+    the sets in partition
+    """
+    partition_union: Set = set()
+    partition_size = 0
+    for s in partition:
+        partition_union = partition_union.union(s)
+        partition_size += len(s)
+
+    return (partition_union == full_set) and (partition_size == len(full_set))
+
+
+class JointMechanism(CentrallyApplicablePrivacyMechanism):
+    """
+    Constructs a new CentrallyApplicablePrivacyMechanism from existing ones.
+    Each existing mechanism is applied to a disjoint subset of the client
+    statistics keys. As such JointMechanism can only be applied to client
+    statistics of type MappedVectorStatistics.
+
+    :param mechanisms_and_keys:
```

Comment:
is it possible to add a feature to also allow specify keys by sub path. e.g. if i have an LLM (which produce many keys) and then some other stats `algo/value1`, `algo/value2`, ... etc i want to put through a mechanism, it would be great if you can achieve this by:

```
mechanisms_and_keys = {
  'algo_mechanism': (m0, ['algo/']),
  'model_mechanism': (m1, [])
}
```

`algo_mechanism` takes all keys which start with `algo/` and `model_mechanism` take all the rest of the keys.

Reply:
Nice, thanks for the suggestion. I've added the feature that allows keys to be matched by subpathing. 

I felt like it was safer not to implement the: [] matches all remaining keys, since this could make it easier to accidentally pass the wrong key to model_mechanism e.g. if I just forget to add a particular key to another mechanism. The way it is currently implemented this would be caught by the check that all keys have been noised. I think this particular model example should be easy to handle in the current implementation by providing list(model.get_parameters().keys()) to mechanisms_and_keys?